### PR TITLE
LUCENE-10168: Only test N-2 codecs on nightly runs.

### PR DESCRIPTION
**Pull request by jpountz**
In order for tests to keep running fast, this annotates all tests of N-2 codecs
with `@Nightly`. To keep good coverage of releases, the smoke tester is now
configured to run nightly tests.

### CODE REVIEW

File: lucene/backward-codecs/src/java/org/apache/lucene/backward_codecs/lucene70/IndexedDISI.java [hunk 1]
```diff
@@ -1,326 +0,0 @@
...
- * </ul>
- *
- * <p>Only ranges that contain at least one value are encoded.
- *
- * <p>This implementation uses 6 bytes per document in the worst-case, which happens in the case
- * that all ranges contain exactly one document.
- *
- * @lucene.internal
- */
-final class IndexedDISI extends DocIdSetIterator {
```
Comment (jpountz) [inline lucene/backward-codecs/src/java/org/apache/lucene/backward_codecs/lucene70/IndexedDISI.java:53]: This was dead code, hence the removal.

**Review by dweiss — APPROVED** [review]

**Review by rmuir — APPROVED** [review]
Thanks for the great cleanup here. You can already see the time improvement at 9.0 by looking at the build time for the test task :) I think it is good solution, the smoketester can take longer, and nightly builds will prevent last-minute surprises.

**Comment by dweiss** [comment]
For what it's worth - we could also make it a dedicated test group (@@Backcompat)... don't know if it makes sense in any kind of isolation from nightly runs though, so perhaps not worth making life more complex.

**Comment by rmuir** [comment]
@dweiss I considered dedicated test group as well, but I like the `@Nightly` solution that keeps it simple and doesn't add more test groups :). 

In fact, we may want to look into GC'ing test groups such as `@BadApple` that lucene is not using. I am also suspicious of the value we get from `@Slow`: they are enabled by default, hence such tests run everywhere by default (PR checks, ordinary jenkins builds), so personally i'd never risk turning them off only to see a code change immediately fail.

On the other hand, `@Nightly` seems appropriate and useful for this situation, most code changes honestly don't impact the index format. If you are making an index format change, it is typically a significant effort anyway, no big deal to just turn on nightly too before making a PR, etc.

**Comment by dweiss** [comment]
I'm used to three levels - all regular tests (for local checks prior to commit), Slow (these run on a CI push) and Nightly (require more resources and/or time). So Slow tests are typically turned off and left for the CI run to verify... which is fine most of the time if it's a pull-request or a branch commit, for example.

But you're right that simpler is usually better and those test groups (other than badapple/awaitsfix) weren't that popular/ used anyway.

**Comment by rmuir** [comment]
I think for most lucene tests currently marked `@Slow`, there are so many paths to fix them to no longer need it:
* Run the test with `tests.profile` and actually fix something slow about it.
* Tone down test's iterations and rely on the "multipliers" to crank it up in jenkins.
* Split test into two methods, a smaller method with less iterations, and a separate bigger Nightly method.
* If it is really just exceptionally crazy slow and nothing can be "toned down", mark it Nightly.

**Comment by dweiss** [comment]
It'd be good to periodically have some kind of sanity check for nightly test runs too as they tend to become a dark, throw-it-under-the-rug place for long-running junk. Even nightly runs should have a reasonable end horizon to complete... We do have the default suite timeout (adjusted on some monster tests) for this but it'd be great to have an over-time trend in the spirit of what Mike has for regression-testing.

**Comment by rmuir** [comment]
Interesting, I agree with that. I personally really do try to avoid just "tossing tests under the rug" to Nightly when working on the tests performance. To me it is the last resort, they are a necessary? evil. I'd rather that we have fast tests organized in a simple system that everyone understands, e.g. no surprises where a PR build fails but things passed for the contributor locally, no surprises where a nightly build fails after a PR build succeeded, etc. 

Often times it is just so easy to fix the tests in another way: less iterations, less threads (only max 2 should ever be needed IMO), less documents, etc etc.

**Pull request merged**